File: DesignPattern/src/09prototype/py/resume.py
# 
import copy

class Resume():
    def __init__(self, name):
        self.name = name
    
    def SetPersonalInfo(self,sex, age):
        self.sex = sex
        self.age = age
    
    def SetWorkExperience(self, timeArea, company):
        self.timeArea = timeArea
        self.company = company

    def Show(self):
        print("{} {} {}".format(self.name, self.sex, self.age))
        print("工作经历: {} {}".format(self.timeArea, self.company))
    
    def Clone(self):
        return copy.copy(self)

if __name__ == "__main__":
    m1 = Resume("Tom")
    m1.SetPersonalInfo("Man", 25)
    m1.SetWorkExperience("xxx-xxx", "Com1")

    m2 = m1.Clone()
    m2.SetPersonalInfo("Man", 30)

    m1.Show()
    m2.Show()

File: DesignPattern/src/09prototype/go/pattern/resume.go
package pattern

import (
	"encoding/json"
	"log"
)

type Resume struct {
	Name string
	Age int
	Sex string
	Work Work
	//TimeArea string
	//Company string
}

func NewResume(name string) (r *Resume) {
	work := NewWork()
	r = &Resume{Name:name, Work:work}
	return
}

func (r *Resume) SetPersonalInfo(sex string, age int) {
	r.Sex = sex
	r.Age = age
}

func (r *Resume) SetWorkInfo(timeArea string, company string) {
	r.Work.TimeArea = timeArea
	r.Work.Company = company
}

func (r *Resume) toString() (ret string){
	bytes, err := json.Marshal(r)
	if err != nil {
		log.Print(err)
	}
	ret = string(bytes)
	return
}

func (r *Resume) Clone() (rc Resume){
	rc.Name = r.Name
	rc.Sex = r.Sex
	rc.Age = r.Age
	//rc.Company = r.Company
	//rc.TimeArea = r.TimeArea
	rc.Work = r.Work
	return
}

